app/core/src/main/java/com/fsck/k9/helper/ListUnsubscribeHelper.kt

```diff
@@ -0,0 +1,63 @@
+package com.fsck.k9.helper
+
+import android.net.Uri
+import com.fsck.k9.mail.Message
+import java.util.regex.Pattern
+
+class ListUnsubscribeHelper {
+    companion object {
+        private const val LIST_UNSUBSCRIBE_HEADER = "List-Unsubscribe"
+        private val MAILTO_CONTAINER_PATTERN = Pattern.compile("<(mailto:.+)>")
+        private val HTTPS_CONTAINER_PATTERN = Pattern.compile("<(https:.+)>")
+
+        // As K-9 Mail is an email client, we prefer a mailto: unsubscribe method
+        // but if none is found, a https URL is acceptable too
+        fun getPreferredListUnsubscribeUri(message: Message?): UnsubscribeUri? {
+            if (message == null) {
+                return null
+            }
+
+            val headerValues = message.getHeader(LIST_UNSUBSCRIBE_HEADER)
+            if (headerValues.isEmpty()) {
+                return null
+            }
+            val listUnsubscribeUris: MutableList<Uri> = ArrayList()
+            for (headerValue in headerValues) {
+                for (headerValuePart in headerValue.split(",")) {
```

Comment:
Nothing is ever easy. URLs can contain commas. So this will fail to correctly extract e.g. this URL:

```eml
List-Unsubscribe: <https://domain.example/one,two>
```

Not splitting on a comma and changing the regular expressions to match non-greedily should be good enough for us.

```kotlin
private val MAILTO_CONTAINER_PATTERN = Pattern.compile("<(mailto:.+?)>")
private val HTTPS_CONTAINER_PATTERN = Pattern.compile("<(https:.+?)>")
```